### PR TITLE
ci: add github actions

### PR DESCRIPTION
**Pull request by mhaber-tophat**
PR moves circleci -> github actions. 

Two workflows. `CI` and `release`. These workflows do the same. 

Pipeline does `service` and `client` builds in parallel then has the release stage wait on them both to finish. 

This tries to do a 1->1 mapping if circleci. Will be making improvements to pipeline in future.  

pipeline:
1. lint
2. build 
3. dry run release/release 

NOTE: To my knowledge theres no way of sharing artifacts between workflows, but id love to not have to redo all the same steps in release workflow. In future probably move client/service stuff to an action so no reused code. Or ensuire semantic release builds the artifacts/ containers

### CODE REVIEW

File: .github/actions/build/action.yml [hunk 1]
```diff
@@ -0,0 +1,15 @@
+name: build
+description: build a sub directory (client or service)
+inputs:
+  project: 
+    description: slient of service
```
Comment (noahnu) [inline .github/actions/build/action.yml:5]: ?

File: .github/actions/detect-node/action.yml [hunk 2]
```diff
@@ -1,5 +1,5 @@
 name: detect-node
```
Comment (noahnu) [inline .github/actions/detect-node/action.yml:1]: (nb) I've renamed this to detect-env in some other projects

File: .github/workflows/pull-request.yml [hunk 3]
```diff
@@ -1,23 +1,74 @@
...
+          docker save -o docker-cache/sanity-docker.tar tophat/sanity-runner:latest
+      - name: Upload artifacts
+        uses: actions/upload-artifact@v2
+        with: 
+          name: client-artifacts
+          path: |
+            client/bin
+            docker-cache
+  service:
+    runs-on: ubuntu-18.04
```
Comment (noahnu) [inline .github/workflows/pull-request.yml:37]: (nb) there's also `ubuntu-20.04`

File: .github/workflows/pull-request.yml [hunk 4]
```diff
@@ -1,23 +1,74 @@
 name: CI
 
-on: [push, pull_request]
+on: [pull_request]
 
 jobs:
-  lint:
+  client:
     runs-on: ubuntu-18.04
     steps:
```
Comment (noahnu) [inline .github/workflows/pull-request.yml:8]: (nb) there's also a cache action you can use for dependencies. Here's an example: https://github.com/tophat/yvm/blob/ee3cde6da3c0ce127453dd41ad2d159b356b05a7/.github/workflows/ci.yml#L27

File: .github/workflows/pull-request.yml [hunk 5]
```diff
@@ -1,23 +1,74 @@
 name: CI
 
-on: [push, pull_request]
+on: [pull_request]
```
Comment (noahnu) [inline .github/workflows/pull-request.yml:3]: (nb) Here's an example of running it for master as well: https://github.com/tophat/yvm/blob/ee3cde6da3c0ce127453dd41ad2d159b356b05a7/.github/workflows/ci.yml#L6

File: .github/workflows/release.yml [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+      - name: build service
+        uses: ./.github/actions/build
+        with:
+          project: service
+      - name: Upload artifacts
+        uses: actions/upload-artifact@v2
+        with: 
+          name: service-artifacts
+          path: service/artifacts/build
+  release:
```
Comment (noahnu) [inline .github/workflows/release.yml:48]: You're missing the deploy website step?

File: .github/workflows/pull-request.yml (hunk 3 above)
Comment (mhaber-tophat) [reply .github/workflows/pull-request.yml:37]: ima just set to `ubuntu-latest` which is still 18.04 but will auto update infuture.

File: .github/actions/build/action.yml (hunk 1 above)
Comment (mhaber-tophat) [reply .github/actions/build/action.yml]: i think i had an aneurysm writing this. i know what i was attempting but wasnt even close.

File: .github/workflows/release.yml (hunk 6 above)
Comment (mhaber-tophat) [reply .github/workflows/release.yml:48]: ahhh ya adding

File: .github/workflows/pull-request.yml (hunk 4 above)
Comment (mhaber-tophat) [reply .github/workflows/pull-request.yml]: added cache

File: .github/workflows/pull-request.yml (hunk 5 above)
Comment (mhaber-tophat) [reply .github/workflows/pull-request.yml]: did this + removed release then. gonna just run release on master branch and not have a seperate release pipeline rn. doesnt make sense

File: .github/workflows/cicd.yml [hunk 7]
```diff
@@ -0,0 +1,172 @@
...
+      - name: install dependencies
+        run: make install
+      - name: package artifacts
+        run: make create-release-package
+      - name: run release dry run
+        run: make deploy-release-dry
+  website:
+    name: Deploy website
+    runs-on: ubuntu-latest
+    if: "!contains(github.event.head_commit.message, 'skip ci')"
```
Comment (noahnu) [inline .github/workflows/cicd.yml:141]: You can add this check to the other jobs as well. The commit from deploying the website doesn't need to run any of the other jobs.

Comment (mhaber-tophat) [reply .github/workflows/cicd.yml:143]: updated

**Review by noahnu — APPROVED** [review]

**Pull request merged**